### PR TITLE
fix(sns/nns): Fix edge case where simultaneous upgrades can cause deadlock

### PR DESCRIPTION
**Pull request by max-dfinity**
In rare cases, when two upgrade proposals are passed at the same time, SNS and NNS root can attempt to execute them both at the same time, causing the canister to become permanently stopped, or to upgrade unsafely (when it is not fully stopped.  

This is fixed by adding a lock to prevent simultaneous upgrade attempts on the same canister.   The second upgrade immediately fails, which can be recovered by making a subsequent proposal if that is desired.  

Queuing and other sequencing mechanisms to try the conflicting upgrade were considered, but the complexity outweighed any benefit, since this is a rare edge case.

### CODE REVIEW

File: rs/nervous_system/lock/src/tests.rs [hunk 1]
```diff
@@ -57,3 +57,99 @@ async fn test_acquire() {
...
+#[tokio::test]
+async fn test_acquire_for_named_locks() {
+    async fn delayed_file_operation(
+        pre_flight_delay_ms: u64,
+        file_path: &str,
+        operation: FileOperation,
+    ) -> bool {
+        sleep(Duration::from_millis(pre_flight_delay_ms)).await;
+        try_file_operation(file_path.to_string(), operation).await
+    }
```
Comment (jasonz-dfinity) [inline rs/nervous_system/lock/src/tests.rs]: It seems this function is duplicate over all the tests for no reason

File: rs/nervous_system/root/src/change_canister.rs [hunk 2]
```diff
@@ -218,13 +220,36 @@ pub struct StopOrStartCanisterRequest {
...
+                canister_id, conflicting_request
+            ));
+        }
+    };
+
+    println!(
+        "{}change_canister: Starting canister change for {}. Request: {:?}",
+        LOG_PREFIX, canister_id, request
+    );
+
```
Comment (jasonz-dfinity) [inline rs/nervous_system/root/src/change_canister.rs]: NCR. This is fine, but not very useful for SNS (since it needs to be `log!`)

File: rs/nervous_system/root/src/change_canister.rs [hunk 3]
```diff
@@ -395,3 +420,113 @@ where
...
+        assert!(result.is_err());
+        let error_msg = result.unwrap_err();
+        assert!(error_msg.contains("currently locked by another change operation"));
+        assert!(error_msg.contains(&format!("{}", canister_id)));
+
+        // Clean up the lock for other tests
+        CANISTER_CHANGE_LOCKS.with(|locks| {
+            locks.borrow_mut().clear();
+        });
+    }
```
Comment (jasonz-dfinity) [inline rs/nervous_system/root/src/change_canister.rs]: I would be surprised if this is needed, because of `thread_local!`?

**Review by jasonz-dfinity — APPROVED** [review]

File: rs/nervous_system/root/src/change_canister.rs (hunk 2 above)
Comment (max-dfinity) [reply rs/nervous_system/root/src/change_canister.rs]: True.  Going to remove, since it's not really high value.

File: rs/nervous_system/lock/src/tests.rs (hunk 1 above)
Comment (max-dfinity) [reply rs/nervous_system/lock/src/tests.rs]: Good point, changed.
